fuel-vm/src/interpreter/executors/main.rs

```diff
@@ -48,46 +50,121 @@ impl<T> Interpreter<PredicateStorage, T> {
-            let mut vm = vm.clone();
+                vm.context = Context::PredicateVerification {
+                    program: predicate.clone(),
+                };
 
-            vm.context = Context::Predicate { program: predicate };
-            vm.set_remaining_gas(remaining_gas);
+                let gas_used = if let Some(x) = input.predicate_gas_used() {
+                    x
+                } else {
+                    return Err(PredicateVerificationFailed::GasNotSpecified);
+                };
+
+                vm.set_gas(gas_used);
 
-            if !matches!(vm.verify_predicate()?, ProgramState::Return(0x01)) {
-                return Err(PredicateVerificationFailed::False);
+                if !matches!(vm.verify_predicate()?, ProgramState::Return(0x01)) {
+                    return Err(PredicateVerificationFailed::False);
+                }
+
+                if vm.registers[RegId::GGAS] != 0 {
+                    return Err(PredicateVerificationFailed::GasMismatch);
+                }
+                cumulative_gas_used = cumulative_gas_used
+                    .checked_add(gas_used)
+                    .ok_or_else(|| PredicateVerificationFailed::OutOfGas)?;
             }
+        }
 
-            remaining_gas = vm.registers[RegId::GGAS];
+        if cumulative_gas_used > checked.transaction().limit() {
+            return Err(PredicateVerificationFailed::OutOfGas);
         }
 
         Ok(PredicatesChecked {
-            gas_used: tx_gas_limit
-                .checked_sub(remaining_gas)
-                .ok_or_else(|| Bug::new(BugId::ID004, GlobalGasUnderflow))?,
+            gas_used: cumulative_gas_used,
         })
     }
+    /// Initialize the VM with the provided transaction and check all predicates defined in the
+    /// inputs. Set the predicate_gas_used to be the actual gas consumed during execution for
+    /// each predicate.
+    ///
+    /// The storage provider is not used since contract opcodes are not allowed for predicates.
+    /// This way, its possible, for the sake of simplicity, it is possible to use
+    /// [unit](https://doc.rust-lang.org/core/primitive.unit.html) as storage provider.
+    ///
+    /// # Debug
+    ///
+    /// This is not a valid entrypoint for debug calls. It will only return a `bool`, and not the
+    /// VM state required to trace the execution steps.
+    pub fn estimate_predicates<Tx>(
+        transaction: &mut Tx,
+        balances: InitialBalances,
+        params: ConsensusParameters,
+        gas_costs: GasCosts,
+    ) -> Result<bool, PredicateVerificationFailed>
+    where
+        Tx: ExecutableTransaction,
+    {
+        let mut vm = Interpreter::with_storage(PredicateStorage::default(), params, gas_costs);
+
+        let tx_gas_limit: u64 = transaction.limit();
+
+        let predicate_gas_limit: u64 = if tx_gas_limit > params.max_gas_per_predicate {
+            params.max_gas_per_predicate
+        } else {
+            tx_gas_limit
+        };
```

Comment:
It is only estimation, not validation. We can use gas limit=)

```suggestion
        let predicate_gas_limit: u64 = params.max_gas_per_predicate;
```

Reply:
i agree, this looks like a useful simplification as the tx_gas_limit could need to be adjusted after estimating predicates